### DISCUSSION TITLE
[v4] CSS Import and Nesting

### DISCUSSION THREAD

**Discussion by dvzrd** [tailwindlabs/tailwindcss#15524] [Help] [answered]
Hello, thank you for releasing v4 beta, I really like the idea of working in css as it makes a lot of things more simpler for reusability than js plugins. Hope you can answer some questions I have related to importing and nesting css

Importing and nesting wasn't a problem in v3 when using `tailwindcss/nesting` with `postcss-preset-env` and `postcss-nesting` plugins. Here's the config I had:

```
const config = {
  plugins: {
    "postcss-import": {},
    "tailwindcss/nesting": {},
    tailwindcss: {},
    "postcss-preset-env": {
      stage: 2,
      minimumVendorImplementations: 2,
      features: { "nesting-rules": false },
    },
  },
};
```
 
In v4, I replaced the above with just the `@tailwindcss/postcss` plugin but my nested styles for some primitive component styles that I was assigning to the components layer were not processed into usable css.

I have a component library with fragmented stylesheets for reusable component styles, which I import into a single components.css file. I then import this and other style collections into a main.css tailwind stylesheet, using the layer function so they can be injected into the correct tailwind layer. Again, not a problem in v3 but for v4 it doesn't work as expected.

So, I have a few questions about import and nesting in v4:

- Is it possible to import two separate theme stylesheets and merge them into a single theme layer/reference - both stylesheets have a @theme layer but I'm unable to merge them together into a single theme when importing them
- Is nesting supposed to work without config now since lightningcss is supposed to handle nested css rule? How does this work with imported stylesheets that don't have tailwind layers inside them, instead the import handles assigning the file to a specific layer.
- Does your app need a main stylesheet or can it build from an external stylesheet, imported into the app's layout as a workspace dependency. Do I have to import it into a stylesheet with a tailwind import instead?

Hoping to get some guidance on this if someone already solved a similar issue with v4 or if the feature is still in development.

I don't have a problem with v4 when using a single stylesheet inside an app but when I have multiple packages with different styles and I want to merge them together with layer imports, things don't exactly work out. The themes aren't injected and the nested styles aren't processed into css.

Thanks for your work on this!

**Comment by wongjn** [comment, marked as answer]
> Is it possible to import two separate theme stylesheets and merge them into a single theme layer/reference - both stylesheets have a `@theme` layer but I'm unable to merge them together into a single theme when importing them

Yes, this should be possible:

```css
/* a.css */
@theme {
  --color-foo: red;
}
```
```css
/* b.css */
@theme {
  --color-bar: blue;
}
```
```css
/* input.css */
@import "tailwindcss";
@import "./a.css";
@import "./b.css";
```
Would compile to something like:
```css
@layer theme, base, components, utilities;
@layer theme {
  :root {
    …
    --color-foo: red;
    --color-bar: blue;
  }
}
```

> Is nesting supposed to work without config now since lightningcss is supposed to handle nested css rule? How does this work with imported stylesheets that don't have tailwind layers inside them, instead the import handles assigning the file to a specific layer.

Nesting works in the sense that Tailwind now uses native CSS nesting.

> Does your app need a main stylesheet or can it build from an external stylesheet, imported into the app's layout as a workspace dependency. 

Tailwind should be able to build from an external stylesheet, `import`ed into the app's layout as a workspace dependency. In v4, the CSS file becomes the defacto configuration file.

> Do I have to import it into a stylesheet with a tailwind import instead?

You shouldn't need to, if the stylesheet itself has `@import "tailwindcss";` in it and the consumer project has Tailwind v4 installed.

**Reply by dvzrd** [reply]
Thanks @wongjn, I managed to merge the themes. I think the issue was using theme layer in the import, didn't need to include that like with the other layer imports because the theme files already have `@theme` so it just works as a regular import.

The only issue I'm having is with some css nesting rules not being compiled. I don't think it's specifically nesting related because some nested rules are working. It's more an issue with some modern css rules that're probably not supported by the current lightningcss presets.

The primary issue is with nesting css classes like this:

```css
.btn {
  &-primary {
    ...
  }
}
```

Seems like the compiler can't read the sass style syntax and only picks up on style rules when defined like this:

```css
.btn {
  &:where(.btn-primary) {
    ...
  }
}
```

Is there a way to configure the preset env for tailwind's lightningcss compiler using the postcss plugin options? Is there plans for a future release to be able to override the compiler options to enable modern css presets?

Thanks again, for answering my questions. Hope you have a happy new year!

**Reply by wongjn** [reply]
> Is there a way to configure the preset env for tailwind's lightningcss compiler using the postcss plugin options?

No.

> Is there plans for a future release to be able to override the compiler options to enable modern css presets?

I'd say that using CSS-native nesting is already pretty modern CSS. The misconception here is that CSS-native nesting does not behave exactly the same as Sass-style nesting.

**Reply by dvzrd** [reply]
Thanks @wongjn, I appreciate your help! The problem is basically syntax sugar so it's not a big deal.

**Reply by dzonatan** [reply]
Just wanted to say that I have faced the same issue with lack of sass-style nesting support in v4. Would be nice to somehow have the ability to opt-in.

**Reply by airtonix** [reply]
Given that native css nesting is not fully supported is there a way we can get the old behaviour without switching back to v3?

https://caniuse.com/css-nesting

**Reply by wongjn** [reply]
It does seem to be supported though, as baseline 2023. Though when building for production, Lightning CSS will flatten the nesting.

**Reply by frontedbe** [reply]
The '&' parent selector in SASS nesting has always been one of my favorite features. It's particularly useful when following BEM methodology, as it eliminates the need to repeatedly type out the parent selector in PCSS files. Guess I'll need to change my workflow and start typing more from now on?

**Reply by wongjn** [reply]
Yes. Though with optimal Tailwind workflow, you shouldn't need to write many BEM classes or any CSS rules at all.

**Reply by frontedbe** [reply]
I might be misunderstanding something. While I really enjoy using Tailwind's utility classes and have become more comfortable with them over time, I find it challenging to maintain readability when multiple classes accumulate in the class attribute. This is especially true when components need different styles in different contexts.
That's why I prefer using @apply in separate .pcss component files - it lets me organize variants and modifiers of components on separate lines in the CSS. For example, when I have a component that uses the same HTML structure but needs a different heading font size in certain cases, I can simply add a modifier class like .component--different-font-size-for-heading { h2 { @apply text-24;} } rather than duplicating the entire component markup just to change the heading's font size. Are there any resources or best practices you could recommend for managing complex component variations more efficiently while maintaining code readability?

**Reply by brittonwalker** [reply]
What I don't fully understand is why this works when running a development build:

```
.post-tease {
  @apply rounded-3xl overflow-hidden border border-black;
  &__card {
    @apply flex flex-col;
  }
  &__title {
    @apply text-gray-900 text-xl font-medium mb-2;
  }
  &__content {
    @apply pt-4 px-4 pb-16;
  }
}
```

But it doesn't compile correctly for a production build.

**Comment by ebnibrahem** [comment]
.project {
    @apply bg-white border-b border-light mb-8 text-primary;
    &-meta {
      @apply flex items-center gap-4 text-gray-500 mb-8;
    }
  }
  
  not woking:
  output:
   .project {...}
    -meta-project {...}

**Reply by jarnesjo** [reply]
What I understand they now using native CSS nesting and concatenation do not work with it [mdn source](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting#concatenation_is_not_possible)

The only thing I have found if you want to create BEM like nesting you have to change approach. 

```
.project {
    @apply bg-white border-b border-light mb-8 text-primary;
    
    &[class*="-meta"] {
        @apply flex items-center gap-4 text-gray-500 mb-8;
    }
}
```

I don't really like it but it seems to be the only way

**Reply by frontedbe** [reply]
Yeah, I find that &[class*="-meta"] {} is even less readable / easy to write than &-meta {}, but since neither approach works for our needs, I'm left with no choice but to write out the complete selectors. While more verbose, this approach ensures maximum clarity and compatibility.

**Reply by brittonwalker** [reply]
Yeah @mdominguez I've had to have a similar approach. This was working in dev:

```
.post-tease {
  @apply rounded-3xl overflow-hidden border border-black;
  &__card {
    @apply flex flex-col;
  }
  &__title {
    @apply text-gray-900 text-xl font-medium mb-2;
  }
  &__content {
    @apply pt-4 px-4 pb-16;
  }
}
```

But for it to work for prod build it would have to be:

```
.post-tease {
  @apply rounded-3xl overflow-hidden border border-black;
    & &__card {
      @apply flex flex-col;
    }
    & &__title {
      @apply text-gray-900 text-xl font-medium mb-2;
    }
    & &__content {
      @apply pt-4 px-4 pb-16;
    }
}
```

I've went with writing out the complete selectors, something I personally feel is not ideal. It has something to do with how lightning css does the build is as far as I've gotten.

**Reply by kamyarkaviani** [reply]
Does lightning CSS even do any processing on top of nested classes? It's likely that they are passed through to the user as nested, and utilizing the browser's capability to deal with CSS nesting instead...

**Comment by lmc3s3f0r** [comment]
Sorry to jump on this thread, but I don't fully understand how to migrate my custom css files from Tailwind v3 to v4.

In my tailwind v3 app I had this file "button.css":

```
.t-button {
  @apply flex select-none border items-center justify-center px-3 py-2 text-sm font-medium !rounded-lg cursor-pointer gap-1.5 transition-colors !duration-normal h-[34px];
  
  & > span[data-pc-section='label'] {
    @apply !duration-0;
  }

  &:disabled {
    @apply pointer-events-none;
  }

  & > .p-badge {
    @apply flex items-center justify-center transition-colors duration-normal text-xs leading-3 px-1.5 py-1 rounded-full;
  }

  &.tb-primary {
    @apply bg-primary-600 text-white border-transparent;

    & > .p-badge {
      @apply text-primary-600 bg-white;
    }

    &:enabled:hover {
      @apply bg-primary-500;
    }

    &:enabled:active {
      @apply bg-primary-600;
    }

    &:disabled {
      @apply bg-secondary-100 text-secondary-400 cursor-default;

      & > .p-badge {
        @apply bg-secondary-400 text-white;
      }
    }
  }

  &.tb-secondary {
    @apply bg-white border-secondary-300 text-secondary-700;

    & > .p-badge {
      @apply text-white bg-primary-600;
    }

    &:enabled:hover {
      @apply text-secondary-800 border-secondary-700 bg-white;
    }

    &:enabled:active {
      @apply text-white border-primary-600 bg-primary-600;

      & > .p-badge {
        @apply bg-white text-primary-600;
      }
    }

    &:disabled {
      @apply bg-secondary-100 text-secondary-400 cursor-default border-transparent;

      & > .p-badge {
        @apply bg-secondary-400 text-white;
      }
    }
  }

  &.p-button-icon-only {
    @apply gap-0;
  }

  .p-button-icon {
    @apply text-lg;
  }
}
```


This was imported in the main css file:
```
@import url('./button.css');

@tailwind base;
@tailwind components;
@tailwind utilities;
```

Now the styles in button.css are not processed, and from your replies I'm unable to figure out what to do. Any hint is greatly appreciated.

**Reply by wongjn** [reply]
Try:
```css
@import "tailwindcss";
@import "./button.css";
```

**Reply by lmc3s3f0r** [reply]
Thanks, it works now. I tried that before but it was throwing an error that I understood now (I was using a custom variable that I hadn't defined yet).

Another small question, hopefully: in v3, I had this in tailwind.config.ts:

```
theme: {
    extend: {
      transitionDuration: {
        short: '150ms',
        normal: '200ms',
        long: '300ms'
      },
    }
  },
```

So then I could use "duration-normal".

How does that translate to v4?

**Reply by wongjn** [reply]
```css
@theme {
  --transition-duration-short: 150ms;
  --transition-duration-normal: 200ms;
  --transition-duration-long: 300ms;
}
```

**Reply by lmc3s3f0r** [reply]
With that, duration-normal will throw an error.

I tried this instead:
duration-(--normal)
duration-(--transition-duration-normal)
duration-[var(--transition-duration-normal)]

But none of them work.

**Reply by wongjn** [reply]
What error does it throw? Seems to be working [in Tailwind Play](https://play.tailwindcss.com/rqJh4rp9np) at least which would suggest something else is causing the error.

**Reply by lmc3s3f0r** [reply]
It says "[plugin:@tailwindcss/vite:generate:serve] Cannot apply unknown utility class `!duration-huge`"

(I defined a "huge" value to make sure it's working).

This is my main.css:
```
@import 'tailwindcss';
@import 'tailwindcss-primeui';
@import './button.css';

@theme {
  --transition-duration-short: 150ms,
  --transition-duration-normal: 200ms,
  --transition-duration-long: 300ms,
  --transition-duration-huge: 3000ms,
}
```

And this is where I use it in button.css:
```
.t-button {
  @apply flex select-none border items-center justify-center px-3 py-2 text-sm font-medium !rounded-lg cursor-pointer gap-1.5 transition-colors !duration-huge h-[34px];
}
```

Tried moving @theme before the import button.css, but same result.

**Reply by wongjn** [reply]
You should terminate each CSS variable declaration with a semicolon (`;`), not a comma (`,`):

```diff
 @theme {
-  --transition-duration-short: 150ms,
-  --transition-duration-normal: 200ms,
-  --transition-duration-long: 300ms,
-  --transition-duration-huge: 3000ms,
+  --transition-duration-short: 150ms;
+  --transition-duration-normal: 200ms;
+  --transition-duration-long: 300ms;
+  --transition-duration-huge: 3000ms;
 }
```

**Reply by lmc3s3f0r** [reply]
Wow, I was trusting either Prettier or my linter would warn me about that ^_^U

Thanks a lot! So far I like what I'm seeing in v4 :)
